File: app/src/main/java/com/passwordmanager/ui/SavePasswordActivity.java
package com.passwordmanager.ui;

import android.os.Bundle;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.WindowCompat;

import com.passwordmanager.R;
import com.passwordmanager.utils.Controller;
import com.passwordmanager.databinding.ActivitySavePasswordBinding;

public class SavePasswordActivity extends AppCompatActivity {
  private Controller controller;
  
  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    ActivitySavePasswordBinding binding = ActivitySavePasswordBinding.inflate(getLayoutInflater());
    setContentView(binding.getRoot());
  
    // Add event onclick listener
    addOnClickListenerOnButton(binding);
    
    // Make window fullscreen
    WindowCompat.setDecorFitsSystemWindows(getWindow(), false);
  }
  
  // Added all the onclick event listiners
  private void addOnClickListenerOnButton(ActivitySavePasswordBinding binding) {
    binding.inputDomain.setOnFocusChangeListener((v, hasFocus) -> {
        if (hasFocus) {
            binding.inputDomain.setHint(R.string.placeholder_domain_field);
        } else {
            binding.inputDomain.setHint("");
        }
    });

    binding.inputUsername.setOnFocusChangeListener((v, hasFocus) -> {
        if (hasFocus) {
            binding.inputUsername.setHint(R.string.placeholder_username_field);
        } else {
            binding.inputUsername.setHint("");
        }
    });

    binding.inputPassword.setOnFocusChangeListener((v, hasFocus) -> {
        if (hasFocus) {
            binding.inputPassword.setHint(R.string.placeholder_password_field);
        } else {
            binding.inputPassword.setHint("");
        }
    });

    binding.savePasswordBtn.setOnClickListener(v -> {
      String domain = binding.inputDomain.getText().toString();
      String username = binding.inputUsername.getText().toString();
      String password = binding.inputPassword.getText().toString();
      String notes = binding.inputNotes.getText().toString();
      
      controller = new Controller(SavePasswordActivity.this);
      int res = controller.savePasswordEntity(domain, username, password, notes);
      
      if (res == -2) {
        Toast.makeText(SavePasswordActivity.this, getString(R.string.warn_fill_form), Toast.LENGTH_SHORT).show();
      } 
      else if (res == -1) {
        Toast.makeText(SavePasswordActivity.this, getString(R.string.fail_msg), Toast.LENGTH_SHORT).show();
      } else {
        Toast.makeText(SavePasswordActivity.this, getString(R.string.sucess_clause) + res, Toast.LENGTH_SHORT).show();
      }
    });
  }
}

File: app/src/main/kotlin/com/jeeldobariya/passcodes/ui/activities/SavePasswordActivity.kt
package com.jeeldobariya.passcodes.ui.activities

import android.content.Context
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.WindowCompat
import androidx.lifecycle.lifecycleScope
import com.jeeldobariya.passcodes.R
import com.jeeldobariya.passcodes.databinding.ActivitySavePasswordBinding
import com.jeeldobariya.passcodes.utils.Controller
import com.jeeldobariya.passcodes.utils.DatabaseOperationException
import com.jeeldobariya.passcodes.utils.InvalidInputException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class SavePasswordActivity : AppCompatActivity() {

    private lateinit var controller: Controller
    private lateinit var binding: ActivitySavePasswordBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        val sharedPrefs = getSharedPreferences(SettingsActivity.THEME_PREFS_NAME, Context.MODE_PRIVATE)
        val savedThemeStyle = sharedPrefs.getInt(SettingsActivity.THEME_KEY, R.style.PasscodesTheme)
        setTheme(savedThemeStyle)
        
        super.onCreate(savedInstanceState)
        binding = ActivitySavePasswordBinding.inflate(layoutInflater)
        setContentView(binding.root)

        controller = Controller(this) // Initialize controller

        // Add event onclick listener
        addOnClickListenerOnButton()

        // Make window fullscreen
        WindowCompat.setDecorFitsSystemWindows(window, false)
    }

    // Added all the onclick event listeners
    private fun addOnClickListenerOnButton() {
        binding.inputDomain.setOnFocusChangeListener { v, hasFocus -> 
            if (hasFocus) {
                binding.inputDomain.setHint(getString(R.string.placeholder_domain_field));
            } else {
                binding.inputDomain.setHint("");
            }
        };

        binding.inputUsername.setOnFocusChangeListener { v, hasFocus -> 
            if (hasFocus) {
                binding.inputUsername.setHint(getString(R.string.placeholder_username_field));
            } else {
                binding.inputUsername.setHint("");
            }
        };

        binding.inputPassword.setOnFocusChangeListener { v, hasFocus -> 
            if (hasFocus) {
                binding.inputPassword.setHint(getString(R.string.placeholder_password_field));
            } else {
                binding.inputPassword.setHint("");
            }
        };

        binding.savePasswordBtn.setOnClickListener {
            val domain = binding.inputDomain.text.toString()
            val username = binding.inputUsername.text.toString()
            val password = binding.inputPassword.text.toString()
            val notes = binding.inputNotes.text.toString()

            performSavePasswordAction(domain, username, password, notes)
        }
    }

    fun performSavePasswordAction(domain: String, username: String, password: String, notes: String) {
        // Launch a coroutine to call the suspend function
        lifecycleScope.launch {
            try {
                val rowId = controller.savePasswordEntity(domain, username, password, notes)
                // Switch back to Main dispatcher for UI updates
                withContext(Dispatchers.Main) {
                    Toast.makeText(this@SavePasswordActivity, "${getString(R.string.success_clause)} $rowId", Toast.LENGTH_SHORT).show()
                    finish()
                }
            } catch (e: InvalidInputException) {
                withContext(Dispatchers.Main) {
                    Toast.makeText(this@SavePasswordActivity, getString(R.string.warn_fill_form), Toast.LENGTH_SHORT).show()
                }
            } catch (e: DatabaseOperationException) {
                withContext(Dispatchers.Main) {
                    Toast.makeText(this@SavePasswordActivity, "${getString(R.string.fail_msg)}: ${e.message}", Toast.LENGTH_LONG).show()
                    e.printStackTrace()
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    Toast.makeText(this@SavePasswordActivity, "${getString(R.string.fail_msg)}: ${e.message}", Toast.LENGTH_LONG).show()
                    e.printStackTrace()
                }
            }
        }
    }
}
